### PR TITLE
Added requirement for codesigning and timestamping certificate

### PR DESCRIPTION
**Pull request by priteshbandi**
Added some requirements for signing certificates and timestamping certificates.

Ref: https://hackmd.io/@pritesh/cert-chain-constraints

### CODE REVIEW

**Comment by gokarnm** [comment]
cc: @shizhMSFT @roywill for review.

File: signature-specification.md [hunk 1]
```diff
@@ -166,26 +166,42 @@ The signing certificate's public key algorithm and size MUST be used to determin
...
+    - For RSA public key, the key length MUST be 2048 bits or higher.
+    - For ECDSA public key, the key length MUST be 256 bits or higher.
+
+#### Other requirements
+
+1. The certificates in the signature MUST be ordered list of X.509 certificate or certificate chain i.e. the certificate containing the public key used to digitally sign the payload must be the first certificate, followed by the intermediate and root certificates in the correct order. This also means
+    - The certificate MUST NOT chain to multiple parents/roots.
+    - The certificate chain MUST NOT contain a certificate that’s not part of the certificate chain.
+1. A Valid certificate chain MUST contain a minimum of two certificates - a leaf and a root certificate.
+1. The certificate in the certificate chain MUST NOT use SHA1WithRSA and ECDSAWithSHA1 signatures.
```
Comment (gokarnm) [inline signature-specification.md]: Nit
```suggestion
1. A valid certificate chain MUST contain a minimum of two certificates - a leaf and a root certificate.
1. Any certificate in the certificate chain MUST NOT use SHA1WithRSA and ECDSAWithSHA1 signatures.
```

File: signature-specification.md [hunk 2]
```diff
@@ -166,26 +166,42 @@ The signing certificate's public key algorithm and size MUST be used to determin
...
+1. Only Basic Constraints, Key Usage, and Extended Key Usage extensions of X.509 certificates are honored. For reset of the extensions, Notary MUST fail open i.e. they MUST NOT be evaluated and honored.
```
Comment (gokarnm) [inline signature-specification.md]:
```suggestion
1. Only Basic Constraints, Key Usage, and Extended Key Usage extensions of X.509 certificates are honored. For rest of the extensions, Notary MUST fail open i.e. they MUST NOT be evaluated or honored.
```

File: signature-specification.md [hunk 3]
```diff
@@ -166,26 +166,42 @@ The signing certificate's public key algorithm and size MUST be used to determin
...
+    - **For timestamping certificate:** If present, the value MUST contain `id-kp-timeStamping` and MUST NOT contain `anyExtendedKeyUsage`, `serverAuth`, `emailProtection` and `codeSigning`.
+1. **Key Length** The certificate MUST abide by the following key length restrictions:
+    - For RSA public key, the key length MUST be 2048 bits or higher.
+    - For ECDSA public key, the key length MUST be 256 bits or higher.
+
+#### Other requirements
+
+1. The certificates in the signature MUST be ordered list of X.509 certificate or certificate chain i.e. the certificate containing the public key used to digitally sign the payload must be the first certificate, followed by the intermediate and root certificates in the correct order. This also means
+    - The certificate MUST NOT chain to multiple parents/roots.
+    - The certificate chain MUST NOT contain a certificate that’s not part of the certificate chain.
```
Comment (roywill) [inline signature-specification.md]: Who is going to check for this condition?

File: signature-specification.md [hunk 4]
```diff
@@ -166,26 +166,42 @@ The signing certificate's public key algorithm and size MUST be used to determin
...
+1. A valid certificate chain MUST contain a minimum of two certificates - a leaf and a root certificate.
+1. Any certificate in the certificate chain MUST NOT use SHA1WithRSA and ECDSAWithSHA1 signatures.
+1. Only Basic Constraints, Key Usage, and Extended Key Usage extensions of X.509 certificates are honored. For rest of the extensions, Notary MUST fail open i.e. they MUST NOT be evaluated or honored.
+1. The certificate in the certificate chain MUST be valid at signing time. Notary MUST NOT enforce validity period nesting for certificates.
```
Comment (roywill) [inline signature-specification.md]: This is weirdly worded?

**Review by roywill — CHANGES_REQUESTED** [review]

**Review by roywill — APPROVED** [review]

File: signature-specification.md (hunk 3 above)
Comment (gokarnm) [reply signature-specification.md]:
```suggestion
    - The certificate chain MUST NOT contain a certificate that’s is unrelated to the certificate chain.
```

File: signature-specification.md [hunk 5]
```diff
@@ -166,26 +166,42 @@ The signing certificate's public key algorithm and size MUST be used to determin
...
-
-- The keyUsage extension MUST be present and MUST be marked critical.
-  The bit positions for digitalSignature MUST be set ([RFC-5280](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.3)).
-- The extKeyUsage extension MUST be present and MUST be marked critical.
-  The value of extension MUST be id-kp-timeStamping ([RFC-5280](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.12)).
-- If the basicConstraints extension is present, the cA field MUST be set false ([RFC-5280](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.9)).
-- The certificate MUST abide by the following key length restrictions:
-  - For RSA public key, the key length MUST be 2048 bits or higher.
-  - For ECDSA public key, the key length MUST be 256 bits or higher.
+The codesigning and timestamping certificates MUST meet the following requirements:
```
Comment (gokarnm) [inline signature-specification.md]:
```suggestion
The codesigning and timestamping certificates MUST meet the following requirements. These requirements are validated both at signature generation time and signature verification time, and are applied to the certificate chain in the signature envelope. These validations are independent of certificate chain validation against a trust store.
```

File: signature-specification.md (hunk 4 above)
Comment (priteshbandi) [reply signature-specification.md]: Reworded it to `The certificates in the certificate chain MUST be valid at signing time. Notary MUST NOT enforce validity period nesting, i.e the validity period for a given certificate may not fall entirely within the validity period of that certificate's issuer certificate.`

**Review by gokarnm — APPROVED** [review]
LGTM! @roywill we reworded sections where you provided feedback. You can check my comments on PR for the suggested reword that was incorporated.

**Review by NiazFK — COMMENTED** [review]
LGTM

**Pull request merged**